cmake: Don't try to convert generator expressions

SebastianBoe opened the pull request:
When generator expressions are passed to zephyr_include_directories it
tries to convert it to an absolute path, but this will result in the
wrong path.

With this change we skip the conversion when a generator expression is
detected and leave it to the user to only use generator expressions
with absolute paths.

This change allows generator expressions to be used.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -79,26 +79,28 @@ endfunction()
+  endforeach()
+endfunction()
+
+function(__maybe_convert_from_relative_to_absolute_path input output)
```
we should not use `_` as prefix for functions.

The `_` is used by CMake to prepend existing functions when they are overloaded, as can be seen by this example:
```
# test.cmake
function(A)
  message("I'm in A")
endfunction()

function(A)
  message("I'm in B")
endfunction()

A()
_A()
```
and running this:
```
$ cmake -P test.cmake
I'm in B
I'm in A
```

Just make a comment and add private to its name, like in this case:
https://github.com/zephyrproject-rtos/zephyr/blob/25e6803705b25602eb0b730a03a254c354bfae9e/cmake/extensions.cmake#L2986-L2987

tejlmand reviewed (changes requested):
The `_` as function name prefix is internally used by CMake, so we shouldn't start using this in function names.

tejlmand commented:
Thanks for fixing this.

SebastianBoe commented:
@tejlmand : Feedback addressed

tejlmand commented:
> @tejlmand : Feedback addressed

Thanks.

Question, why are we converting the relative paths in the first place ?
`target_include_directories()` already supports relative paths:
https://cmake.org/cmake/help/v3.22/command/target_include_directories.html
> Specified include directories may be absolute paths or relative paths.

so we can just remove all the check and modification of relative paths and pass all arguments directly to the native `target_include_directories()` function, and thus we avoid checking both for generator expressions and relative paths.

SebastianBoe commented:
I can't remember why we started doing it.

SebastianBoe commented:
The relative path is relative to the target isn't it? But we want it relative to the invocation.

tejlmand commented:
> The relative path is relative to the target isn't it? But we want it relative to the invocation.

relative to the current source dir as far as I'm aware, but good point, worth double checking.

However, I assume the reason it might have been done like this is because of `target_sources()` originally didn't support relative paths (support for relative paths was added in CMake 3.13).
And thus removed similar code in `zephyr_sources()` with this commit:
https://github.com/zephyrproject-rtos/zephyr/commit/e217c1b172965adf151d3e9abc71a0b56a78ed97

I think we should check this, and remove the relative path handling if possible.

SebastianBoe commented:
Out-of-scope for this PR IMHO, but I agree, if we can simplify then that is great :)

tejlmand commented:
> Out-of-scope for this PR IMHO, but I agree, if we can simplify then that is great :)

I don't thinks so, if this is the scope of the PR:
> This change allows generator expressions to be used.

and that can be achieved by removing relative path conversion instead of adding more logic, then that is within scope of this PR, imho.

https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html
> The target_include_directories() command accepts both relative and absolute include directories:
> ....
> Relative paths are interpreted relative to the source directory where the command appears.

So please remove existing logic instead to fix this, instead of adding more.

SebastianBoe committed:
cmake: Don't translate from relative to absolute include paths

Don't translate from relative to absolute include paths.

It is lost to history why this was necessary and it is causing
problems with generator expressions so we remove the logic.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
Removed the translation.

I'm a bit sceptical that this will be OK.

Anyone extracting include paths will run into trouble, but let's see what CI says.

tejlmand commented:
> Anyone extracting include paths will run into trouble

Why ?
```
target_include_directories(app PUBLIC inc/rel_test)
add_custom_target(test_rel_inc COMMAND ${CMAKE_COMMAND} -E echo \"$<TARGET_PROPERTY:app,INCLUDE_DIRECTORIES>\")
```

results in:
```
$ ninja test_rel_inc
/projects/github/ncs/zephyr/samples/hello_world/inc/rel_test;/projects/github/ncs/zephyr/include;...
```

properly expanded.

tejlmand reviewed (dismissed):
lgtm

SebastianBoe commented:
I meant the evaluation of generator expressions at configure-time, but I see we only do this for the COMPILE_LANGUAGE generator expression, so perhaps this is safe.

SebastianBoe commented:
@tejlmand : +1 needs to be re-applied

tejlmand reviewed (approved):
"lgtm - reapproved"

tejlmand reviewed (approved):
lgtm - reapproved
